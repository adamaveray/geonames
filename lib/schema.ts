export interface FieldTypes {
  string: string;
  int: number;
  float: number;
  array: string[];
  date: Date;
}

export type Schema = typeof schema;

const schema = {
  geonameId: { type: 'int', nullable: false },
  name: { type: 'string', nullable: false },
  asciiName: { type: 'string', nullable: true },
  alternateNames: { type: 'array', nullable: true },
  latitude: { type: 'float', nullable: false },
  longitude: { type: 'float', nullable: false },
  featureClass: { type: 'string', nullable: false },
  featureCode: { type: 'string', nullable: false },
  countryCode: { type: 'string', nullable: false },
  cc2: { type: 'string', nullable: true },
  admin1Code: { type: 'string', nullable: true },
  admin2Code: { type: 'string', nullable: true },
  admin3Code: { type: 'string', nullable: true },
  admin4Code: { type: 'string', nullable: true },
  population: { type: 'int', nullable: false },
  elevation: { type: 'int', nullable: true },
  dem: { type: 'int', nullable: false },
  timezone: { type: 'string', nullable: false },
  modificationDate: { type: 'date', nullable: false },
} as const satisfies Record<string, { type: keyof FieldTypes; nullable: boolean }>;

export default schema;
