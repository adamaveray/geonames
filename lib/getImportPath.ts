import { dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

export default function getImportPath(importUrl: string, basename: string): string {
  return `${dirname(fileURLToPath(importUrl))}/${basename}`;
}
