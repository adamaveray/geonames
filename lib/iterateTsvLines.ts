import * as fs from 'node:fs';

const LF = '\n'.charCodeAt(0);
const TAB = '\t'.charCodeAt(0);
export default function* iterateTsvLines(pathname: string, encoding: BufferEncoding): Generator<string[], void> {
  const fd = fs.openSync(pathname, 'r');

  const byteBuffer = Buffer.alloc(1);
  let fragmentBytes: number[] = [];
  let lineColumns: string[] = [];

  const step = (): number => fs.readSync(fd, byteBuffer, 0, 1, null);

  const endColumn = () => {
    lineColumns.push(Buffer.from(fragmentBytes).toString(encoding));
    fragmentBytes = [];
  };

  while (true) {
    // Read next byte
    if (step() === 0) {
      // End of file
      break;
    }

    const byte = byteBuffer[0];
    if (byte === TAB) {
      // End of column
      endColumn();
      continue;
    }

    if (byte === LF) {
      // End of line
      endColumn();
      if (lineColumns.length > 0) {
        yield lineColumns;
        lineColumns = [];
      }

      continue;
    }

    // Regular byte
    if (byte != null) {
      fragmentBytes.push(byte);
    }
  }

  // Output incomplete line
  if (fragmentBytes.length > 0) {
    endColumn();
  }
  if (lineColumns.length > 0) {
    yield lineColumns;
  }
}
