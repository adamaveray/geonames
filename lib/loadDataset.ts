import iterateTsvLines from './iterateTsvLines';
import schema from './schema';
import type { FieldTypes, Schema } from './schema';

export type DataGenerator = Generator<Row, void>;

type Row = {
  [key in keyof Schema]: Schema[key]['nullable'] extends true
    ? FieldTypes[Schema[key]['type']] | undefined
    : FieldTypes[Schema[key]['type']];
};

const casters: { [key in keyof FieldTypes]: (value: string) => FieldTypes[key] | undefined } = {
  string: (value) => value,
  int: (value) => parseInt(value, 10),
  float: (value) => parseFloat(value),
  array: (value) => value.split(','),
  date: (value) => new Date(`${value}T00:00:00Z`),
};

export default function* loadDataset(pathname: string): DataGenerator {
  const parseField = (field: keyof Schema, value: string): Row[typeof field] => {
    const { type, nullable } = schema[field];
    if (value === '') {
      if (!nullable) {
        throw new Error(`Field "${String(field)}" cannot be empty`);
      }
      return (type === 'array' ? [] : undefined) as Row[typeof field];
    }
    return casters[type](value);
  };

  for (const cols of iterateTsvLines(pathname, 'utf8')) {
    yield Object.fromEntries(
      Object.keys(schema).map((field, index) => [field, parseField(field as keyof Schema, cols[index] ?? '')]),
    ) as Row;
  }
}
