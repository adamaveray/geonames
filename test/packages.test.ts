import cities500 from '../packages/cities-500-structured';
import cities1000 from '../packages/cities-1000-structured';
import cities5000 from '../packages/cities-5000-structured';
import cities15000 from '../packages/cities-15000-structured';
import schema from '../lib/schema';
import type { FieldTypes } from '../lib/schema';
import { DataGenerator } from '../lib/loadDataset';

const getFieldsOfType = (assertedType: keyof FieldTypes): (keyof typeof schema)[] =>
  Object.entries(schema)
    .filter(([, { type }]) => type === assertedType)
    .map(([field]) => field as keyof typeof schema);

const stringFields = getFieldsOfType('string');
const intFields = getFieldsOfType('int');
const floatFields = getFieldsOfType('float');
const arrayFields = getFieldsOfType('array');
const dateFields = getFieldsOfType('date');

const expectType = (field: keyof typeof schema, value: any, isValid: boolean): void => {
  const { nullable } = schema[field];
  expect(isValid || (nullable && value == null)).toBeTruthy();
};

describe('packages', () => {
  const schemaKeys = Object.keys(schema);

  it.each<[key: string, builder: () => DataGenerator]>([
    ['cities-15000', cities15000],
    ['cities-5000', cities5000],
    ['cities-1000', cities1000],
    ['cities-500', cities500],
  ])('provides correct output', (key, builder) => {
    const firstRecord = builder().next().value;
    expect(firstRecord).toMatchSnapshot(key);
  });

  it.each<[key: string, builder: () => DataGenerator]>([
    ['cities-15000', cities15000],
    ['cities-5000', cities5000],
    ['cities-1000', cities1000],
    ['cities-500', cities500],
  ])('load rows correctly', (_, builder) => {
    let i = 0;
    const limit = 50; // Should be representative enough
    for (const record of builder()) {
      expect(Object.keys(record)).toEqual(schemaKeys);

      for (const field of stringFields) {
        expectType(field, record[field], typeof record[field] === 'string');
      }
      for (const field of intFields) {
        expectType(field, record[field], typeof record[field] === 'number');
      }
      for (const field of floatFields) {
        expectType(field, record[field], typeof record[field] === 'number');
      }
      for (const field of arrayFields) {
        expectType(field, record[field], Array.isArray(record[field]));
      }
      for (const field of dateFields) {
        expectType(field, record[field], record[field] instanceof Date);
      }

      i++;
      if (i >= limit) {
        break;
      }
    }
  });
});
