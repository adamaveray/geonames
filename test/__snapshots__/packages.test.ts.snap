// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packages provides correct output: cities-500 1`] = `
{
  "admin1Code": "02",
  "admin2Code": undefined,
  "admin3Code": undefined,
  "admin4Code": undefined,
  "alternateNames": [],
  "asciiName": "Soldeu",
  "cc2": undefined,
  "countryCode": "AD",
  "dem": 1832,
  "elevation": undefined,
  "featureClass": "P",
  "featureCode": "PPL",
  "geonameId": 3038999,
  "latitude": 42.57688,
  "longitude": 1.66769,
  "modificationDate": 2017-11-06T00:00:00.000Z,
  "name": "Soldeu",
  "population": 602,
  "timezone": "Europe/Andorra",
}
`;

exports[`packages provides correct output: cities-1000 1`] = `
{
  "admin1Code": "02",
  "admin2Code": undefined,
  "admin3Code": undefined,
  "admin4Code": undefined,
  "alternateNames": [
    "Ehl Tarter",
    "Эл Тартер",
  ],
  "asciiName": "El Tarter",
  "cc2": undefined,
  "countryCode": "AD",
  "dem": 1721,
  "elevation": undefined,
  "featureClass": "P",
  "featureCode": "PPL",
  "geonameId": 3039154,
  "latitude": 42.57952,
  "longitude": 1.65362,
  "modificationDate": 2012-11-03T00:00:00.000Z,
  "name": "El Tarter",
  "population": 1052,
  "timezone": "Europe/Andorra",
}
`;

exports[`packages provides correct output: cities-5000 1`] = `
{
  "admin1Code": "06",
  "admin2Code": undefined,
  "admin3Code": undefined,
  "admin4Code": undefined,
  "alternateNames": [
    "San Julia",
    "San Julià",
    "Sant Julia de Loria",
    "Sant Julià de Lòria",
    "Sant-Zhulija-de-Lorija",
    "sheng hu li ya-de luo li ya",
    "Сант-Жулия-де-Лория",
    "サン・ジュリア・デ・ロリア教区",
    "圣胡利娅-德洛里亚",
    "圣胡利娅－德洛里亚",
  ],
  "asciiName": "Sant Julia de Loria",
  "cc2": undefined,
  "countryCode": "AD",
  "dem": 921,
  "elevation": undefined,
  "featureClass": "P",
  "featureCode": "PPLA",
  "geonameId": 3039163,
  "latitude": 42.46372,
  "longitude": 1.49129,
  "modificationDate": 2013-11-23T00:00:00.000Z,
  "name": "Sant Julià de Lòria",
  "population": 8022,
  "timezone": "Europe/Andorra",
}
`;

exports[`packages provides correct output: cities-15000 1`] = `
{
  "admin1Code": "08",
  "admin2Code": undefined,
  "admin3Code": undefined,
  "admin4Code": undefined,
  "alternateNames": [
    "Ehskal'des-Ehndzhordani",
    "Escaldes",
    "Escaldes-Engordany",
    "Les Escaldes",
    "esukarudesu=engorudani jiao qu",
    "lai sai si ka er de-en ge er da",
    "Эскальдес-Энджордани",
    "エスカルデス＝エンゴルダニ教区",
    "萊塞斯卡爾德-恩戈爾達",
    "萊塞斯卡爾德－恩戈爾達",
  ],
  "asciiName": "les Escaldes",
  "cc2": undefined,
  "countryCode": "AD",
  "dem": 1033,
  "elevation": undefined,
  "featureClass": "P",
  "featureCode": "PPLA",
  "geonameId": 3040051,
  "latitude": 42.50729,
  "longitude": 1.53414,
  "modificationDate": 2008-10-15T00:00:00.000Z,
  "name": "les Escaldes",
  "population": 15853,
  "timezone": "Europe/Andorra",
}
`;
